src/particles.cpp

```diff
@@ -281,6 +281,58 @@ ParticleSpawner::ParticleSpawner(IGameDef *gamedef, LocalPlayer *player,
 	}
 }
 
+void ParticleSpawner::spawnParticle(ClientEnvironment *env, float radius,
+	bool is_attached, const v3f &attached_pos, float attached_yaw)
+{
+	v3f ppos = m_player->getPosition() / BS;
+	v3f pos = random_v3f(m_minpos, m_maxpos);
+
+	// Need to apply this first or the following check
+	// will be wrong for attached spawners
+	if (is_attached) {
+		pos.rotateXZBy(attached_yaw);
+		pos += attached_pos;
+	}
+
+	if (pos.getDistanceFrom(ppos) > radius)
+		return;
+
+	v3f vel = random_v3f(m_minvel, m_maxvel);
+	v3f acc = random_v3f(m_minacc, m_maxacc);
+
+	if (is_attached) {
+		// Apply attachment yaw
+		vel.rotateXZBy(attached_yaw);
+		acc.rotateXZBy(attached_yaw);
+	}
+
+	float exptime = rand() / (float)RAND_MAX
+			* (m_maxexptime - m_minexptime)
+			+ m_minexptime;
```

Comment:
don't we have a macro to do this?

Reply:
`linearInterpolation` in `noise.cpp` does that but I couldn't find a macro or static function.